java/ql/lib/semmle/code/java/Type.qll

```diff
@@ -37,6 +37,16 @@ predicate hasSubtype(RefType t, Type sub) {
   typeVarSubtypeBound(t, sub) and t != sub
 }
 
+/**
+ * Holds if reference type `anc` is a direct or indirect supertype of `sub`, including itself.
+ */
+cached
+predicate hasDescendant(RefType anc, Type sub) {
+  anc = sub
+  or
+  exists(RefType mid | hasSubtype(anc, mid) and hasDescendant(mid, sub))
```

Comment:
Usage of `hasSubtype` has its issues with generic and parameterized types, most notably:
- `getAnAcestor()` never gets for a parameterized type (e.g. `java.util.List<String>`) the generic source declaration (e.g. `java.util.List`)
- `hasSubtype` seems to be unable to return subtypes for generic types, e.g. for `java.util.List` it does not get `java.util.ArrayList` ([example query](https://lgtm.com/query/5617555852827933522/)); maybe that is actually a bug? (can report that separately if you want)

By adding more predicates which suffer from these issues, it will become more likely that users will use these predicates by accident for generic or parameterized types and run into false negatives or not working queries. Though maybe not considering generic and parameterized types here is acceptable / desired for performance?
I just wanted to point that out, in case you were not aware of it.

Related to this might be #5521 and #5595.

Reply:
I take your point about people being more likely to use these predicates if we make it easier for them, but I don't think the new predicates are much easier to use than e.g. `hasSubtype*`. So as this PR doesn't make the behaviour worse, I think I'll go ahead and merge it.

The `List`/`ArrayList` example does look suspicious at first glance, though. Could you go ahead and file a ticket for that please, so we can take a proper look into it.

Reply:
Have created #8339 for this.